src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs

```diff
@@ -65,34 +67,63 @@ internal static void Return(this ExportTraceServiceRequest request)
             return;
         }
 
-        foreach (var scope in resourceSpans.ScopeSpans)
+        foreach (var scopeSpan in resourceSpans.ScopeSpans)
         {
-            scope.Spans.Clear();
-            SpanListPool.Add(scope);
+            scopeSpan.Spans.Clear();
+            SpanListPool.Add(scopeSpan);
         }
     }
 
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    internal static ScopeSpans GetSpanListFromPool(string name, string? version)
+    internal static ScopeSpans GetSpanListFromPool(ActivitySource activitySource, int maxTags, int? attributeValueLengthLimit)
     {
-        if (!SpanListPool.TryTake(out var spans))
+        if (!SpanListPool.TryTake(out var scopeSpans))
         {
-            spans = new ScopeSpans
+            scopeSpans = new ScopeSpans
             {
                 Scope = new InstrumentationScope
                 {
-                    Name = name, // Name is enforced to not be null, but it can be empty.
-                    Version = version ?? string.Empty, // NRE throw by proto
+                    Name = activitySource.Name, // Name is enforced to not be null, but it can be empty.
+                    Version = activitySource.Version ?? string.Empty, // NRE throw by proto
                 },
             };
-        }
-        else
-        {
-            spans.Scope.Name = name;
-            spans.Scope.Version = version ?? string.Empty;
+
+            if (activitySource.Tags != null)
+            {
+                var scopeAttributes = scopeSpans.Scope.Attributes;
+
+                if (activitySource.Tags is IReadOnlyList<KeyValuePair<string, object?>> activitySourceTagsList)
+                {
+                    for (int i = 0; i < activitySourceTagsList.Count; i++)
+                    {
+                        if (scopeAttributes.Count < maxTags)
+                        {
+                            OtlpTagWriter.Instance.TryWriteTag(ref scopeAttributes, activitySourceTagsList[i], attributeValueLengthLimit);
+                        }
+                        else
+                        {
+                            scopeSpans.Scope.DroppedAttributesCount++;
+                        }
+                    }
+                }
+                else
+                {
+                    foreach (var tag in activitySource.Tags ?? [])
```

Comment:
There is a `null` check around the whole thing now so probably don't need the coalesce.

```suggestion
                    foreach (var tag in activitySource.Tags)
```